src/code/sched.c

```diff
@@ -165,90 +227,145 @@ OSScTask* func_800C89D4(SchedContext* sc, OSScTask* task) {
+    s32 nextState = state;
     OSScTask* gfxTask = sc->gfxListHead;
     OSScTask* audioTask = sc->audioListHead;
 
-    if (sc->doAudio && (ret & OS_SC_SP)) {
-        *sp = audioTask;
-        ret &= ~OS_SC_SP;
-        sc->doAudio = 0;
+    if (sc->doAudio && (state & OS_SC_SP)) {
+        // Audio Task, RSP is available
+
+        // Return next audio task
+        *spTaskOut = audioTask;
+        // RSP required
+        nextState &= ~OS_SC_SP;
+        //! @bug If there is more than one audio task in the queue at any time, unsetting doAudio here
+        //! will cause only one task to be processed until a new audio task is enqueued. In practice, audio
+        //! tasks are sent infrequently enough that there are never two audio tasks in the queue.
+        sc->doAudio = false;
+        // Advance task queue
         sc->audioListHead = sc->audioListHead->next;
         if (sc->audioListHead == NULL) {
             sc->audioListTail = NULL;
         }
     } else if (gfxTask != NULL) {
-        if (gfxTask->state & OS_SC_YIELDED || !(gfxTask->flags & OS_SC_NEEDS_RDP)) {
-            if (ret & OS_SC_SP) {
-                *sp = gfxTask;
-                ret &= ~OS_SC_SP;
+        // GFX Task
+
+        if ((gfxTask->state & OS_SC_YIELDED) || !(gfxTask->flags & OS_SC_NEEDS_RDP)) {
+            // If this is a yielded GFX task, or the RDP is not needed for this GFX task
+
+            if (state & OS_SC_SP) {
+                // If the RSP is available, return next graphics task
+
+                *spTaskOut = gfxTask;
+                // RSP required
+                nextState &= ~OS_SC_SP;
+                // Advance task queue
                 sc->gfxListHead = sc->gfxListHead->next;
                 if (sc->gfxListHead == NULL) {
                     sc->gfxListTail = NULL;
                 }
             }
-        } else if (ret == (OS_SC_SP | OS_SC_DP)) {
-            if (gfxTask->framebuffer == NULL || func_800C89D4(sc, gfxTask) != NULL) {
-                *sp = *dp = gfxTask;
-                ret &= ~(OS_SC_SP | OS_SC_DP);
+        } else if (state == (OS_SC_SP | OS_SC_DP)) {
+            // Both the RSP and RDP are available, check requested framebuffer
+            if (gfxTask->framebuffer == NULL || Sched_GfxTaskFramebufferValid(sc, gfxTask) != NULL) {
+                // Return next graphics task
+                *spTaskOut = *dpTaskOut = gfxTask;
+                // RSP and RDP both required
+                nextState &= ~(OS_SC_SP | OS_SC_DP);
+                // Advance task queue
                 sc->gfxListHead = sc->gfxListHead->next;
                 if (sc->gfxListHead == NULL) {
                     sc->gfxListTail = NULL;
                 }
             }
         }
     }
-    return ret;
+    return nextState;
 }
 
-void func_800C8BC4(SchedContext* sc, OSScTask* task) {
+/**
+ * Routine to swap the framebuffer on task completion
+ */
+void Sched_TaskSwapBuffer(SchedContext* sc, OSScTask* task) {
```

Comment:
I feel like this is misleading since it will usually not swap the framebuffer instantly, but most likely from `pendingSwapBuf1` in `Sched_HandleRetrace`

I don't see how to address this in the name without ending up with a too long name

Reply:
`Sched_TaskRegisterBufferSwap` ? Or request? Although it's not exactly a request, it can't exactly be declined

Reply:
`TaskPendingBuffer`? `SetPendingBuffer`?

I'm also not sure what the word `Task` is doing there, is it a verb (to task) or a noun (`OSScTask`)

At the very least the comment on the function should mention the framebuffer is "pending" and not necessarily swapped immediately